File: src/graph.cpp
#include "graph.h"


namespace shortestpaths {


/*
 * Construct a new empty graph.
 */
Graph::Graph() {
  this->adjacency = std::vector<std::vector<std::tuple<int, int>>>();;
}


/*
 * Deconstruct a graph.
 */
Graph::~Graph() {

}


/*
 * Add a vertex, returning its index.
 */
int Graph::add_vertex() {
  this->adjacency.push_back(std::vector<std::tuple<int, int>>());
  return this->adjacency.size() - 1;
}


/*
 * Add an edge with weight w between vertices v and u.
 */
void Graph::add_edge(int v, int u, int w) {
  this->adjacency[v].push_back(std::make_tuple(u, w));
  this->adjacency[u].push_back(std::make_tuple(v, w));
}


/*
 * Returns the weight of the edge between vertices v and u.
 */
int Graph::get_weight(int v, int u) {
  for (auto edge : this->adjacency[v]) {
    int i; int w;
    std::tie(i, w) = edge;

    if (i == u) {
      return w;
    }
  }
}


}; // namespace shortestpaths

File: src/main/java/Graph.java
package tobfeo.splib;

import java.util.ArrayList;
import java.lang.IllegalArgumentException;
import java.lang.Integer;

public class Graph {
  /**
   * A simple graph, implemented as an adjacency list.
   */

  public ArrayList<ArrayList<Pair<Integer, Integer>>> adjacency;

  public Graph() {
    this.adjacency = new ArrayList<ArrayList<Pair<Integer, Integer>>>();
  }

  public Integer addVertex()
  {
    this.adjacency.add(new ArrayList<>());
    return this.adjacency.size() - 1;
  }

  public void addEdge(Integer v, Integer u, Integer w)
  {
    this.adjacency.get(v).add(new Pair<>(u, w));
    this.adjacency.get(u).add(new Pair<>(v, w));
  }

  public void remove(Integer v)
  {
    this.adjacency.remove(v);
  }

  public Integer getWeight(Integer v, Integer u)
  {
    for (Pair<Integer, Integer> edge : this.adjacency.get(v))
    {
      if (edge.x == u)
      {
        return edge.y;
      }
    }
    return null;
  }

}
